### PR TITLE
[ROCm][Bugfix] Fix init-time bias dtype cast when gate.out_dtype is None

### PR DESCRIPTION
**Pull request by rbrugaro-amd**
## Purpose

Fix a regression on ROCm where the init-time `e_score_correction_bias` dtype cast introduced in #39999 became a no-op after #39141 removed the `set_out_dtype()` call, re-introducing a per-layer elementwise kernel on every forward pass.

## Root Cause

PR #39999 (merged Apr 24) moved the bias dtype conversion from runtime to model construction:

```python
self.gate.e_score_correction_bias.data = (
    self.gate.e_score_correction_bias.data.to(self.gate.out_dtype)
)
```

This relied on `self.gate.out_dtype` being set by an explicit `set_out_dtype()` call that existed immediately above:

```python
self.gate.set_out_dtype(
    torch.float32
    if self.experts.quant_method.is_monolithic
    and self.experts.routing_method_type == RoutingMethodType.DeepSeekV3
    else torch.bfloat16
)
```

Three days later, PR #39141 (commit `c8bbe0518`, Apr 27) removed that `set_out_dtype()` call as part of a TRTLLM routing refactor, leaving `self.gate.out_dtype = None` for all DeepSeek/Kimi-K2 models.

Since `tensor.to(None)` is a no-op in PyTorch, the bias stays fp32 while the gate output (router logits) is bf16. The conditional runtime safety net in `_aiter_ops.py` then fires on every forward pass:

```python
if correction_bias.dtype != gating_output.dtype:
    correction_bias = correction_bias.to(gating_output.dtype)
```

This launches the same elementwise cast kernel per MoE layer that #39999 was designed to eliminate.

## Fix

Fall back to `self.gate.weight.dtype` when `out_dtype` is `None`:

```python
gate_out_dtype = self.gate.out_dtype or self.gate.weight.dtype
```

This is safe because the entire code block is gated by `self.is_rocm_aiter_moe_enabled`, and on ROCm the gate always uses Tier-3 dispatch (F.linear via ReplicatedLinear), where the output dtype equals the weight dtype (bf16). If `out_dtype` is properly set in the future, it takes precedence over the fallback.

## Tests:
Elementwise no longer present in Kimi-K2-thinking-MXFP4 MI355x trace

Accuracy PASS
openai-completions ({'model': '/data/workloads-inference/models/Kimi-K2.6-MXFP4', 'base_url': 'http://0.0.0.0:8888/v1/completions', 'add_bos_token': True, 'enforce_eager': True, 'num_concurrent': 64, 'max_retries': 10, 'max_gen_toks': 1024, 'tokenizer_backend': 'huggingface'}), gen_kwargs: ({}), limit: 250.0, num_fewshot: 5, batch_size: 64
|Tasks|Version|     Filter     |n-shot|  Metric   |   |Value|   |Stderr|
|-----|------:|----------------|-----:|-----------|---|----:|---|-----:|
|gsm8k|      3|flexible-extract|     5|exact_match|↑  | 0.94|±  |0.0151|
|     |       |strict-match    |     5|exact_match|↑  | 0.94|±  |0.0151|

## Test Plan

- [X] Verify the elementwise kernel is no longer present in traces for Kimi-K2-thinking-MXFP4 MI355x
- [X] Accuracy: GSM8K 5-shot (score >= 0.90)

## References

- #39999 — original fix that moved the cast to init-time
- #39141 (commit `c8bbe0518`) — removed `set_out_dtype()`, causing this regression

Made with [Cursor](https://cursor.com)

### CODE REVIEW

**Review by Rohan138 — APPROVED** [review]
LGTM, but @bnellnm could you PTAL?

**Review by bnellnm — APPROVED** [review]

**Review by gshtras — APPROVED** [review]
